src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs

```diff
@@ -101,15 +118,31 @@ private MemoryCacheEntryOptions GetMemoryCacheEntryOptions(string relativePath)
             return options;
         }
 
-        private static string NormalizePath(string path)
+        private string GetNormalizedPath(string relativePath)
         {
-            // We need to allow for scenarios where the application was precompiled on a machine with forward slashes
-            // but is being run in one with backslashes (or vice versa). To this effect, we'll normalize paths to
-            // use backslashes for lookups and storage in the dictionary.
-            path = path.Replace('/', '\\');
-            path = path.TrimStart('\\');
+            Debug.Assert(relativePath != null);
+            if (relativePath.Length == 0)
+            {
+                return relativePath;
+            }
 
-            return path;
+            lock (_cacheKeyLookupLock)
+            {
+                string normalizedPath;
+                if (!_normalizedPathLookup.TryGetValue(relativePath, out normalizedPath))
```

Comment:
curious, why a lock rather than a ConcurrentDictionary? Lock seems more error prone and you're only protecting one resource